### ISSUE TITLE
Fermatas

### ISSUE THREAD

**Issue by joeberkovitz** [w3c/mnx#106]
This issue notes the need for fermatas.

MusicXML nests fermatas inside other things like note or barline elements. In MNX-Common, the model could perhaps be simpler, making a fermata a simple direction whose location determines what it goes over or under.

Also note that fermatas can appear above a staff, below a staff or both. MusicXML allows a fermata to occur up to two times in various contexts to cover this. In MNX-Common, we could consider handling this question with a style property.

**Comment by mogenslundholm** [comment]
Shall fermatas be present in all parts? I think they should, but some notewriters writes only one (in the first part). Therefore I have a special handling if a fermata is missing (add it if another part has a fermata).  But the fermata need not be at the same place. What do you think?

**Comment by joeberkovitz** [comment]
I think fermatas do need to be present in all parts, because in non-homophonic music the fermatas may be placed on different notes in different parts. (For example, see various Bach chorales or Behind Bars pp. 188-189.)

**Comment by joeberkovitz** [comment]
Let me clarify that by saying, MNX-Common should not _require_ that fermatas be present in all parts. I am only saying that fermatas need not agree in the various parts, it is up to the composer or engraver to supply a fermata in each part if that is what is desired.

**Comment by adrianholovaty** [comment]
We chatted about fermatas in the most recent MNX meeting, and here's a concrete proposal:

We add a fermata object, with the following keys (all optional):

* **"duration"**: specifies the duration of a fermata's pause. These values are intentionally open-ended, as opposed to being precise amounts of time. We expect consuming software will have their own internal mappings of how long a "normal" fermata is versus a "long" one, etc. Valid values: "auto" (software is free to calculate on its own), "none" (fermata doesn't affect playback), "veryLong", "long", "normal", "short", "veryShort". Default if unspecified is "auto".
* **"symbol"**: specifies the visual symbol to use. This is deliberately not a SMuFL glyph name because that would be a bit too low-level (there are multiple glyphs for the same "symbol", such as the upside-down version of each fermata gyph). Valid values: the same as [MusicXML](https://w3c-cg.github.io/musicxml/musicxml-reference/data-types/fermata-shape/), minus the empty string option. Default if unspecified is "normal".
* **"orient"**: specifies the orientation. Valid values: "above", "below" (upside down) and "auto". Default if unspecified is "auto".

Fermata objects can live in three places:

* Within an event, in a new "fermata" key. This represents a fermata that's rhythmically aligned with the event.
* Within a global measure, in a new "fermata" key. This represents a fermata that's attached to the end of the measure (traditionally rendered over the barline).
* On a full measure rest object, in a new "fermata" key.

Comments and suggestions welcome!

**Comment by adrianholovaty** [comment]
@rpatters1 Thanks for the catch there. I've just updated my proposal above to include that as a third option. Everything else look good?

**Comment by rpatters1** [comment]
I agree with the idea of a more abstract symbol than glyph. But I wonder if as a general issue we need to grapple with glyphs as a consistent option for any symbol (fermata, marking, dynamic, clef, etc.) This is perhaps out of scope for the fermata per se.

I do not like the name `duration` because it conflicts with the idea of duration as a notation, e.g., `event.duration`. Perhaps the name should imply that it is a playback-only feature. `playbackLength` or something. (Lol, something better than that.)

Should we not use naming consistency with markings? That would be `pointing` rather than `orient`. (Or vice versa.) Generally, `orient` is either undocumented and/or disappearing in this spec.

BTW: I am getting 404 now for any of the links to the online reference.

**Comment by rpatters1** [comment]
Thinking more about `orient`/`pointing` in the context of fermatas on barlines, they are rendered over the barline of single-staff parts, but more commonly rendered both over the top and and under the bottom barline on measure stacks in scores. Sometimes the pattern is repeated for each group barline. Are we trying to capture/encode this?

**Comment by samuelbradshaw** [comment]
What's the reasoning behind open-ended rather than precise duration values?

**Comment by joeberkovitz** [comment]
> What's the reasoning behind open-ended rather than precise duration values?

Without taking a stand on this choice, I'd like to point out that there is a potential need across the board for flexible duration indications that are either absolute (milliseconds, ticks, etc) or subjective ("long", "veryLong" and so on).

With respect to the previous comment, a request for absolute duration overrides for events is already noted in #27. If that issue gets addressed via some mechanism, it could answer the need for more precise fermata durations, either through application to the fermata itself or to the underlying event or object that the fermata modifies (which may be more generally useful).

It's also an interesting question as to whether there is a parallel issue to #27: should any event be allowed to be modified with _subjective_ durations (rather than absolute, as in the issue)?

**Comment by mscuthbert** [comment]
> What's the reasoning behind open-ended rather than precise duration values?

One reason is that several formats shapes imply longer or shorter durations that are still not precise, but not always with as standardized as the SMuFL glyph names imply. 

We discussed also--but punted for now--text markings affecting playback length such as "lunga"

**Comment by adrianholovaty** [comment]
I've added fermata support to the spec in 34593d4c538d. Here's the new documentation:

https://w3c-cg.github.io/mnx/docs/mnx-reference/objects/fermata/

Regarding `orient` versus `pointing`: the approach here is to specify the orientation (whether the glyph is above or below the staff, or auto). I've seen published music that uses a right-side-up fermata below the staff, so it seems we could solve that with a separate `pointing` attribute.

For multi-instrument music that puts a fermata above the top staff and below the bottom staff (and not in the inner staves), I expect we'll handle that via a style attribute or an attribute of layouts.

Also, we didn't explicitly discuss this, so I wanted to document the thinking: I didn't put fermata into `event.markings` because it feels like a "heavier" concept than simply a marking. But this is pretty subjective, and I could see a valid argument being made for putting it into `event.markings`.

**Comment by rpatters1** [comment]
For myself, I appreciate that fermata is *not* part of `event.marking`. None of the other markings can be attached to measure rests or barlines. It is definitely its own thing from my perspective.

I would like to see `pointing` vs `orient` clarified in `event.markings` as well, though. Specifically, above/below/auto (relative to the `event`) vs the glyph direction. It seems like most of the `event.markings` should have `orient`. Meanwhile, I still don't understand the purpose of `pointing` on a plain accent.

**Comment by rpatters1** [comment]
I just looked at the document page. The fermata symbols are using hyphenated names, but everywhere else in the MNX spec uses camel case. Is this just a case of a copy/paste error from musicxml?

**Comment by adrianholovaty** [comment]
@rpatters1 Thanks for the feedback! I've just done some followup work:

* 1cc1beaae925ce88c022561084fb615216865af4 — removed `pointing` from accent objects
* 2ba4be1ba6b3e30eb52efd02bcfb819b51e2ebbe — properly documented `strong-accent.pointing`
* e6706b211fdd71073a605770790dde778f652e68 — added `fermata.pointing`
* 9d9d3fc0403acad04c90c3e11cc42defad9828f3 — changed the fermata symbol options to be camelCased
* 4868cce9dd42ab8e0c64aee7fbabb010f14f9e5c — added `orient` to all event markings
* a141ee373f086c516642b40e5ad073b0eef05fe6 — documented the enum values for the orientation object

With that, I'll go ahead and mark this as closed. There are still open questions about how to encode the following...

* More granular fermata playback details
* Omission of fermata glyphs from certain staves in a system

...but I think those would be best considered in a more holistic design for playback and styling.
